<a href="https://colab.research.google.com/github/Tendo4/Prediction-of-Product-Sales/blob/main/Tendo_Elisha_Project_1_final_core.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Mount google drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# Import packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
pd.set_option('display.max_columns',100)
import missingno
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler, OneHotEncoder, OrdinalEncoder
from sklearn.pipeline import make_pipeline
from sklearn.impute import SimpleImputer
from sklearn.ensemble import RandomForestRegressor
# Set pandas as the default output for sklearn
from sklearn import set_config
set_config(transform_output='pandas')

In [ ]:
# Load data directly from url
path='/content/drive/MyDrive/CodingDojo/01-Fundamentals/Week02/Data/sales_predictions_2023.csv'
df = pd.read_csv(path)
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

##Data Cleaning

In [ ]:
#checking for duplicates
duplicated_rows=df.duplicated()
duplicated_rows.sum()

0

In [ ]:
##datatypes
df.dtypes

Item_Identifier               object
Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
Item_Outlet_Sales            float64
dtype: object

In [ ]:
#identifying missing values
null=df.isna()
null_sum=null.sum()
null_sum

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

In [ ]:
# Use .select_dtypes to view only object columns
cat_cols=df.select_dtypes(include='object')
cat_cols

Item_Identifier Item_Fat_Content              Item_Type  \
0              FDA15          Low Fat                  Dairy   
1              DRC01          Regular            Soft Drinks   
2              FDN15          Low Fat                   Meat   
3              FDX07          Regular  Fruits and Vegetables   
4              NCD19          Low Fat              Household   
...              ...              ...                    ...   
8518           FDF22          Low Fat            Snack Foods   
8519           FDS36          Regular           Baking Goods   
8520           NCJ29          Low Fat     Health and Hygiene   
8521           FDN46          Regular            Snack Foods   
8522           DRG01          Low Fat            Soft Drinks   

     Outlet_Identifier Outlet_Size Outlet_Location_Type        Outlet_Type  
0               OUT049      Medium               Tier 1  Supermarket Type1  
1               OUT018      Medium               Tier 3  Supermarket Type2  
2               OUT049      Medium               Tier 1  Supermarket Type1  
3               OUT010         NaN               Tier 3      Grocery Store  
4               OUT013        High               Tier 3  Supermarket Type1  
...                ...         ...                  ...                ...  
8518            OUT013        High               Tier 3  Supermarket Type1  
8519            OUT045         NaN               Tier 2  Supermarket Type1  
8520            OUT035       Small               Tier 2  Supermarket Type1  
8521            OUT018      Medium               Tier 3  Supermarket Type2  
8522            OUT046       Small               Tier 1  Supermarket Type1  

[8523 rows x 7 columns]

In [ ]:
# Check the nunique for just the object cols
cat_cols.nunique()

Item_Identifier         1559
Item_Fat_Content           5
Item_Type                 16
Outlet_Identifier         10
Outlet_Size                3
Outlet_Location_Type       3
Outlet_Type                4
dtype: int64

- Note that out of the 7 categorical columns remaining:
  - 6 columns have a small number of categories  
  - 1 column has over 100 categories (Item Identifier).

- We do not want to keep categorical features with so many categories.


- **Drop the Item Identifier column.**

In [ ]:
# Drop the columns listed above
d=cat_cols.drop(['Item_Identifier'],axis=1)
d

Item_Fat_Content              Item_Type Outlet_Identifier Outlet_Size  \
0             Low Fat                  Dairy            OUT049      Medium   
1             Regular            Soft Drinks            OUT018      Medium   
2             Low Fat                   Meat            OUT049      Medium   
3             Regular  Fruits and Vegetables            OUT010         NaN   
4             Low Fat              Household            OUT013        High   
...               ...                    ...               ...         ...   
8518          Low Fat            Snack Foods            OUT013        High   
8519          Regular           Baking Goods            OUT045         NaN   
8520          Low Fat     Health and Hygiene            OUT035       Small   
8521          Regular            Snack Foods            OUT018      Medium   
8522          Low Fat            Soft Drinks            OUT046       Small   

     Outlet_Location_Type        Outlet_Type  
0                  Tier 1  Supermarket Type1  
1                  Tier 3  Supermarket Type2  
2                  Tier 1  Supermarket Type1  
3                  Tier 3      Grocery Store  
4                  Tier 3  Supermarket Type1  
...                   ...                ...  
8518               Tier 3  Supermarket Type1  
8519               Tier 2  Supermarket Type1  
8520               Tier 2  Supermarket Type1  
8521               Tier 3  Supermarket Type2  
8522               Tier 1  Supermarket Type1  

[8523 rows x 6 columns]

In [ ]:
for k in d:
  print(f"Value Counts for {k}")
  print(df[k].value_counts())
  print('\n')


Value Counts for Item_Fat_Content
Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64


Value Counts for Item_Type
Fruits and Vegetables    1232
Snack Foods              1200
Household                 910
Frozen Foods              856
Dairy                     682
Canned                    649
Baking Goods              648
Health and Hygiene        520
Soft Drinks               445
Meat                      425
Breads                    251
Hard Drinks               214
Others                    169
Starchy Foods             148
Breakfast                 110
Seafood                    64
Name: Item_Type, dtype: int64


Value Counts for Outlet_Identifier
OUT027    935
OUT013    932
OUT049    930
OUT046    930
OUT035    930
OUT045    929
OUT018    928
OUT017    926
OUT010    555
OUT019    528
Name: Outlet_Identifier, dtype: int64


Value Counts for Outlet_Size
Medium    2793
Small     2388
High       932
Name: Outlet_Size, 

### Checking & Fixing Consistency: Item Fat Content

In [ ]:
# Check for inconsistencies by inspecting the value_counts for Item Fat Content

df['Item_Fat_Content'].value_counts()

Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64

In [ ]:
# Replace "low fat and LF" with "Low Fat" and 'reg' with 'Regular'
df['Item_Fat_Content'] = df['Item_Fat_Content'].replace({'low fat':"Low Fat",'LF':'Low Fat','reg':'Regular'})
# Check the value counts again to confirm
df['Item_Fat_Content'].value_counts()

Low Fat    5517
Regular    3006
Name: Item_Fat_Content, dtype: int64

In [ ]:
# Define features (X) and target (y).
y = df['Item_Outlet_Sales']
X = df.drop(columns='Item_Outlet_Sales')

In [ ]:
## Train test split
X_train, X_test, y_train, y_test = train_test_split(X,y,random_state=42)
X_train.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
4776           NCG06       16.350          Low Fat         0.029565   
7510           FDV57       15.250          Regular         0.000000   
5828           FDM27       12.350          Regular         0.158716   
5327           FDG24        7.975          Low Fat         0.014628   
4810           FDD05       19.350          Low Fat         0.016645   

         Item_Type  Item_MRP Outlet_Identifier  Outlet_Establishment_Year  \
4776     Household  256.4646            OUT018                       2009   
7510   Snack Foods  179.7660            OUT018                       2009   
5828          Meat  157.2946            OUT049                       1999   
5327  Baking Goods   82.3250            OUT035                       2004   
4810  Frozen Foods  120.9098            OUT045                       2002   

     Outlet_Size Outlet_Location_Type        Outlet_Type  
4776      Medium               Tier 3  Supermarket Type2  
7510      Medium               Tier 3  Supermarket Type2  
5828      Medium               Tier 1  Supermarket Type1  
5327       Small               Tier 2  Supermarket Type1  
4810         NaN               Tier 2  Supermarket Type1

### Define groups of features using lists:



In [ ]:
#categorical columns
#NB : we had already dropped the Item identifier feature so we assigned categorical columns to d

categorical_cols = X_train.select_dtypes('object').columns
categorical_cols

Index(['Item_Identifier', 'Item_Fat_Content', 'Item_Type', 'Outlet_Identifier',
       'Outlet_Size', 'Outlet_Location_Type', 'Outlet_Type'],
      dtype='object')

In [ ]:
# Drop the Item Identifier columns
m=categorical_cols.drop(['Item_Identifier'])
m

Index(['Item_Fat_Content', 'Item_Type', 'Outlet_Identifier', 'Outlet_Size',
       'Outlet_Location_Type', 'Outlet_Type'],
      dtype='object')

In [ ]:
#numerical columns
numeric_cols = X_train.select_dtypes('number').columns
numeric_cols

Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
       'Outlet_Establishment_Year'],
      dtype='object')

###One Hot Encoder for Categorical columns

####imputing categorical columns

In [ ]:
# Instantaite and fit the imputer for categorical features
impute_missing = SimpleImputer(strategy='constant', fill_value='MISSING')
impute_missing.fit(X_train[m])

SimpleImputer(fill_value='MISSING', strategy='constant')

In [ ]:
# Transform the training and test categorical features and save as new vars
X_train_cat_imputed = impute_missing.transform(X_train[m])
X_test_cat_imputed = impute_missing.transform(X_test[m])

# confirm no null values remaining
X_train_cat_imputed.isna().sum()

Item_Fat_Content        0
Item_Type               0
Outlet_Identifier       0
Outlet_Size             0
Outlet_Location_Type    0
Outlet_Type             0
dtype: int64

In [ ]:
# Instantiate one hot encoder
ohe_encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore')

In [ ]:
# Fit the OneHotEncoder on the training data
ohe_encoder.fit(X_train_cat_imputed)

OneHotEncoder(handle_unknown='ignore', sparse_output=False)

In [ ]:
# Transform the training data
X_train_cat_encoded = ohe_encoder.transform(X_train_cat_imputed)
X_train_cat_encoded.head()

Item_Fat_Content_Low Fat  Item_Fat_Content_Regular  \
4776                       1.0                       0.0   
7510                       0.0                       1.0   
5828                       0.0                       1.0   
5327                       1.0                       0.0   
4810                       1.0                       0.0   

      Item_Type_Baking Goods  Item_Type_Breads  Item_Type_Breakfast  \
4776                     0.0               0.0                  0.0   
7510                     0.0               0.0                  0.0   
5828                     0.0               0.0                  0.0   
5327                     1.0               0.0                  0.0   
4810                     0.0               0.0                  0.0   

      Item_Type_Canned  Item_Type_Dairy  Item_Type_Frozen Foods  \
4776               0.0              0.0                     0.0   
7510               0.0              0.0                     0.0   
5828               0.0              0.0                     0.0   
5327               0.0              0.0                     0.0   
4810               0.0              0.0                     1.0   

      Item_Type_Fruits and Vegetables  Item_Type_Hard Drinks  \
4776                              0.0                    0.0   
7510                              0.0                    0.0   
5828                              0.0                    0.0   
5327                              0.0                    0.0   
4810                              0.0                    0.0   

      Item_Type_Health and Hygiene  Item_Type_Household  Item_Type_Meat  \
4776                           0.0                  1.0             0.0   
7510                           0.0                  0.0             0.0   
5828                           0.0                  0.0             1.0   
5327                           0.0                  0.0             0.0   
4810                           0.0                  0.0             0.0   

      Item_Type_Others  Item_Type_Seafood  Item_Type_Snack Foods  \
4776               0.0                0.0                    0.0   
7510               0.0                0.0                    1.0   
5828               0.0                0.0                    0.0   
5327               0.0                0.0                    0.0   
4810               0.0                0.0                    0.0   

      Item_Type_Soft Drinks  Item_Type_Starchy Foods  \
4776                    0.0                      0.0   
7510                    0.0                      0.0   
5828                    0.0                      0.0   
5327                    0.0                      0.0   
4810                    0.0                      0.0   

      Outlet_Identifier_OUT010  Outlet_Identifier_OUT013  \
4776                       0.0                       0.0   
7510                       0.0                       0.0   
5828                       0.0                       0.0   
5327                       0.0                       0.0   
4810                       0.0                       0.0   

      Outlet_Identifier_OUT017  Outlet_Identifier_OUT018  \
4776                       0.0                       1.0   
7510                       0.0                       1.0   
5828                       0.0                       0.0   
5327                       0.0                       0.0   
4810                       0.0                       0.0   

      Outlet_Identifier_OUT019  Outlet_Identifier_OUT027  \
4776                       0.0                       0.0   
7510                       0.0                       0.0   
5828                       0.0                       0.0   
5327                       0.0                       0.0   
4810                       0.0                       0.0   

      Outlet_Identifier_OUT035  Outlet_Identifier_OUT045  \
4776                       0.0                       0.0   
7510                       0.0

In [ ]:
# Transform the testing data
X_test_cat_encoded = ohe_encoder.transform(X_test_cat_imputed)
X_test_cat_encoded.head()

Item_Fat_Content_Low Fat  Item_Fat_Content_Regular  \
7503                       1.0                       0.0   
2957                       1.0                       0.0   
7031                       0.0                       1.0   
1084                       0.0                       1.0   
856                        0.0                       1.0   

      Item_Type_Baking Goods  Item_Type_Breads  Item_Type_Breakfast  \
7503                     0.0               0.0                  0.0   
2957                     0.0               0.0                  0.0   
7031                     0.0               0.0                  0.0   
1084                     0.0               0.0                  0.0   
856                      0.0               0.0                  0.0   

      Item_Type_Canned  Item_Type_Dairy  Item_Type_Frozen Foods  \
7503               0.0              0.0                     1.0   
2957               0.0              0.0                     0.0   
7031               1.0              0.0                     0.0   
1084               0.0              0.0                     0.0   
856                0.0              0.0                     0.0   

      Item_Type_Fruits and Vegetables  Item_Type_Hard Drinks  \
7503                              0.0                    0.0   
2957                              0.0                    0.0   
7031                              0.0                    0.0   
1084                              0.0                    0.0   
856                               0.0                    0.0   

      Item_Type_Health and Hygiene  Item_Type_Household  Item_Type_Meat  \
7503                           0.0                  0.0             0.0   
2957                           1.0                  0.0             0.0   
7031                           0.0                  0.0             0.0   
1084                           0.0                  0.0             0.0   
856                            0.0                  0.0             1.0   

      Item_Type_Others  Item_Type_Seafood  Item_Type_Snack Foods  \
7503               0.0                0.0                    0.0   
2957               0.0                0.0                    0.0   
7031               0.0                0.0                    0.0   
1084               0.0                0.0                    0.0   
856                0.0                0.0                    0.0   

      Item_Type_Soft Drinks  Item_Type_Starchy Foods  \
7503                    0.0                      0.0   
2957                    0.0                      0.0   
7031                    0.0                      0.0   
1084                    1.0                      0.0   
856                     0.0                      0.0   

      Outlet_Identifier_OUT010  Outlet_Identifier_OUT013  \
7503                       0.0                       1.0   
2957                       0.0                       0.0   
7031                       0.0                       0.0   
1084                       0.0                       0.0   
856                        0.0                       0.0   

      Outlet_Identifier_OUT017  Outlet_Identifier_OUT018  \
7503                       0.0                       0.0   
2957                       0.0                       0.0   
7031                       0.0                       0.0   
1084                       0.0                       0.0   
856                        0.0                       0.0   

      Outlet_Identifier_OUT019  Outlet_Identifier_OUT027  \
7503                       0.0                       0.0   
2957                       0.0                       0.0   
7031                       0.0                       0.0   
1084                       0.0                       1.0   
856                        0.0                       0.0   

      Outlet_Identifier_OUT035  Outlet_Identifier_OUT045  \
7503                       0.0                       0.0   
2957                       0.0

###Standard scaler for numeric columns

In [ ]:
# Instantaite and fit the imputer for numeric features
impute_median = SimpleImputer(strategy='median')
impute_median.fit(X_train[numeric_cols])

SimpleImputer(strategy='median')

In [ ]:
# Transform the training and test numeric features and save as new vars
X_train_num_imputed = impute_median.transform(X_train[numeric_cols])
X_test_num_imputed = impute_median.transform(X_test[numeric_cols])

# confirm no null values remaining
X_train_num_imputed.isna().sum()

Item_Weight                  0
Item_Visibility              0
Item_MRP                     0
Outlet_Establishment_Year    0
dtype: int64

In [ ]:
# instantiate standard scaler and fit on training data
scaler = StandardScaler()
scaler.fit(X_train_num_imputed)

StandardScaler()

In [ ]:
# Apply the scaling transformation
X_train_num_scaled = scaler.transform(X_train_num_imputed)
X_test_num_scaled = scaler.transform(X_test_num_imputed)
X_train_num_scaled.head()

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year
4776     0.827485        -0.712775  1.828109                   1.327849
7510     0.566644        -1.291052  0.603369                   1.327849
5828    -0.121028         1.813319  0.244541                   0.136187
5327    -1.158464        -1.004931 -0.952591                   0.732018
4810     1.538870        -0.965484 -0.336460                   0.493686

In [ ]:
# check describe() for pre-scaling data
X_train_num_imputed.describe().round(2)

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year
count      6392.00          6392.00   6392.00                    6392.00
mean         12.86             0.07    141.98                    1997.86
std           4.22             0.05     62.63                       8.39
min           4.56             0.00     31.29                    1985.00
25%           9.50             0.03     94.15                    1987.00
50%          12.65             0.05    144.11                    1999.00
75%          16.10             0.09    186.90                    2004.00
max          21.35             0.33    266.89                    2009.00

In [ ]:
X_train_num_scaled.describe().round(2)

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year
count      6392.00          6392.00   6392.00                    6392.00
mean         -0.00            -0.00      0.00                      -0.00
std           1.00             1.00      1.00                       1.00
min          -1.97            -1.29     -1.77                      -1.53
25%          -0.80            -0.76     -0.76                      -1.29
50%          -0.05            -0.23      0.03                       0.14
75%           0.77             0.56      0.72                       0.73
max           2.01             5.13      1.99                       1.33

##Creating tuples

###Categorical and numeric pipelines

In [ ]:
# Prepare Categorical pipeline


impute_missing = SimpleImputer(strategy='constant',fill_value='MISSING')
ohe_encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore')
ohe_pipe = make_pipeline(impute_missing, ohe_encoder)
ohe_pipe

Pipeline(steps=[('simpleimputer',
                 SimpleImputer(fill_value='MISSING', strategy='constant')),
                ('onehotencoder',
                 OneHotEncoder(handle_unknown='ignore', sparse_output=False))])

In [ ]:
# Prepare numeric piepline
num_cols = X_train.select_dtypes('number').columns

mean_imputer = SimpleImputer(strategy="mean")
scaler = StandardScaler()

#Numeric Pipeline
numeric_pipe = make_pipeline(mean_imputer, scaler)
numeric_pipe

Pipeline(steps=[('simpleimputer', SimpleImputer()),
                ('standardscaler', StandardScaler())])

###Numerical and catregorical tuples

In [ ]:
# Making a numeric tuple for ColumnTransformer
num_tuple = ('numeric', numeric_pipe, num_cols)
num_tuple

('numeric',
 Pipeline(steps=[('simpleimputer', SimpleImputer()),
                 ('standardscaler', StandardScaler())]),
 Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
        'Outlet_Establishment_Year'],
       dtype='object'))

In [ ]:
# Making a ohe_tuple for ColumnTransformer
ohe_tuple = ('categorical', ohe_pipe, m)
ohe_tuple

('categorical',
 Pipeline(steps=[('simpleimputer',
                  SimpleImputer(fill_value='MISSING', strategy='constant')),
                 ('onehotencoder',
                  OneHotEncoder(handle_unknown='ignore', sparse_output=False))]),
 Index(['Item_Fat_Content', 'Item_Type', 'Outlet_Identifier', 'Outlet_Size',
        'Outlet_Location_Type', 'Outlet_Type'],
       dtype='object'))

In [ ]:
# Instantiate with verbose_feature_names_out=False
col_transformer = ColumnTransformer([num_tuple, ohe_tuple],
                                    verbose_feature_names_out=False)
col_transformer

ColumnTransformer(transformers=[('numeric',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer()),
                                                 ('standardscaler',
                                                  StandardScaler())]),
                                 Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
       'Outlet_Establishment_Year'],
      dtype='object')),
                                ('categorical',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer(fill_value='MISSING',
                                                                strategy='constant')),
                                                 ('onehotencoder',
                                                  OneHotEncoder(handle_unknown='ignore',
                                                                sparse_output=False))]),
                                 Index(['Item_Fat_Content', 'Item_Type', 'Outlet_Identifier', 'Outlet_Size',
       'Outlet_Location_Type', 'Outlet_Type'],
      dtype='object'))],
                  verbose_feature_names_out=False)

In [ ]:
# Fit on training data
col_transformer.fit(X_train)

ColumnTransformer(transformers=[('numeric',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer()),
                                                 ('standardscaler',
                                                  StandardScaler())]),
                                 Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
       'Outlet_Establishment_Year'],
      dtype='object')),
                                ('categorical',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer(fill_value='MISSING',
                                                                strategy='constant')),
                                                 ('onehotencoder',
                                                  OneHotEncoder(handle_unknown='ignore',
                                                                sparse_output=False))]),
                                 Index(['Item_Fat_Content', 'Item_Type', 'Outlet_Identifier', 'Outlet_Size',
       'Outlet_Location_Type', 'Outlet_Type'],
      dtype='object'))],
                  verbose_feature_names_out=False)

#####Transforming training and testing data

In [ ]:
# Transform the training data
X_train_processed = col_transformer.transform(X_train)
# Transform the testing data
X_test_processed = col_transformer.transform(X_test)
# View the processed training data
X_train_processed.head()

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year  \
4776     0.817249        -0.712775  1.828109                   1.327849   
7510     0.556340        -1.291052  0.603369                   1.327849   
5828    -0.131512         1.813319  0.244541                   0.136187   
5327    -1.169219        -1.004931 -0.952591                   0.732018   
4810     1.528819        -0.965484 -0.336460                   0.493686   

      Item_Fat_Content_Low Fat  Item_Fat_Content_Regular  \
4776                       1.0                       0.0   
7510                       0.0                       1.0   
5828                       0.0                       1.0   
5327                       1.0                       0.0   
4810                       1.0                       0.0   

      Item_Type_Baking Goods  Item_Type_Breads  Item_Type_Breakfast  \
4776                     0.0               0.0                  0.0   
7510                     0.0               0.0                  0.0   
5828                     0.0               0.0                  0.0   
5327                     1.0               0.0                  0.0   
4810                     0.0               0.0                  0.0   

      Item_Type_Canned  Item_Type_Dairy  Item_Type_Frozen Foods  \
4776               0.0              0.0                     0.0   
7510               0.0              0.0                     0.0   
5828               0.0              0.0                     0.0   
5327               0.0              0.0                     0.0   
4810               0.0              0.0                     1.0   

      Item_Type_Fruits and Vegetables  Item_Type_Hard Drinks  \
4776                              0.0                    0.0   
7510                              0.0                    0.0   
5828                              0.0                    0.0   
5327                              0.0                    0.0   
4810                              0.0                    0.0   

      Item_Type_Health and Hygiene  Item_Type_Household  Item_Type_Meat  \
4776                           0.0                  1.0             0.0   
7510                           0.0                  0.0             0.0   
5828                           0.0                  0.0             1.0   
5327                           0.0                  0.0             0.0   
4810                           0.0                  0.0             0.0   

      Item_Type_Others  Item_Type_Seafood  Item_Type_Snack Foods  \
4776               0.0                0.0                    0.0   
7510               0.0                0.0                    1.0   
5828               0.0                0.0                    0.0   
5327               0.0                0.0                    0.0   
4810               0.0                0.0                    0.0   

      Item_Type_Soft Drinks  Item_Type_Starchy Foods  \
4776                    0.0                      0.0   
7510                    0.0                      0.0   
5828                    0.0                      0.0   
5327                    0.0                      0.0   
4810                    0.0                      0.0   

      Outlet_Identifier_OUT010  Outlet_Identifier_OUT013  \
4776                       0.0                       0.0   
7510                       0.0                       0.0   
5828                       0.0                       0.0   
5327                       0.0                       0.0   
4810                       0.0                       0.0   

      Outlet_Identifier_OUT017  Outlet_Identifier_OUT018  \
4776                       0.0                       1.0   
7510                       0.0                       1.0   
5828                       0.0                       0.0   
5327                       0.0                       0.0   
4810                       0.0                       0.0   

      Outlet_Identifier_OUT019  Outlet_Identifier_OUT027  \


#Project 1 final starts Here

In [ ]:
#Instantiate the model
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg

LinearRegression()

In [ ]:
# Fit the model on the training data
lin_reg.fit(X_train_processed, y_train)

LinearRegression()

In [ ]:
# Get predictions for the training data
y_predictions_train = lin_reg.predict(X_train_processed)
# Get predictions for the testing data
y_predictions_test = lin_reg.predict(X_test_processed)

##Evaluation using regression metrics

In [ ]:
# Calculating MAE with numpy
train_MAE = np.mean(np.abs(y_predictions_train - y_train))
test_MAE = np.mean(np.abs(y_predictions_test - y_test))
print(f'Model Training MAE: {train_MAE:,.2f}')
print(f'Model Testing MAE: {test_MAE:,.2f}')

Model Training MAE: 847.13
Model Testing MAE: 804.13


In [ ]:
# Calculating MSE with numpy
train_MSE = np.mean((y_predictions_train - y_train)**2)
test_MSE = np.mean((y_predictions_test - y_test)**2)
print(f'Model Training MSE: {train_MSE:,.2f}')
print(f'Model Testing MSE: {test_MSE:,.2f}')

Model Training MSE: 1,297,560.41
Model Testing MSE: 1,194,366.48


In [ ]:
# Calculating RMSE with numpy
train_RMSE = np.sqrt(np.mean((y_predictions_train - y_train)**2))
test_RMSE = np.sqrt(np.mean((y_predictions_test - y_test)**2))
print(f'Model Training RMSE: {train_RMSE:,.2f}')
print(f'Model Testing RMSE: {test_RMSE:,.2f}')

Model Training RMSE: 1,139.11
Model Testing RMSE: 1,092.87


In [ ]:
# Calculating R-Squared
train_r2 = np.corrcoef(y_train, y_predictions_train)[0][1]**2
test_r2 = np.corrcoef(y_test, y_predictions_test)[0][1]**2
print(f'Model Training R2: {train_r2:.2f}')
print(f'Model Testing R2: {test_r2:.2f}')

Model Training R2: 0.56
Model Testing R2: 0.57


#####The testing model has a higher R squared value therefore its a better model than the training model

##Random Forests

In [ ]:
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
def regression_metrics(y_true, y_pred, label='', verbose = True, output_dict=False):
  # Get metrics
  mae = mean_absolute_error(y_true, y_pred)
  mse = mean_squared_error(y_true, y_pred)
  rmse = mean_squared_error(y_true, y_pred, squared=False)
  r_squared = r2_score(y_true, y_pred)
  if verbose == True:
    # Print Result with Label and Header
    header = "-"*60
    print(header, f"Regression Metrics: {label}", header, sep='\n')
    print(f"- MAE = {mae:,.3f}")
    print(f"- MSE = {mse:,.3f}")
    print(f"- RMSE = {rmse:,.3f}")
    print(f"- R^2 = {r_squared:,.3f}")
  if output_dict == True:
      metrics = {'Label':label, 'MAE':mae,
                 'MSE':mse, 'RMSE':rmse, 'R^2':r_squared}
      return metrics
def evaluate_regression(reg, X_train, y_train, X_test, y_test, verbose = True,
                        output_frame=False):
  # Get predictions for training data
  y_train_pred = reg.predict(X_train)
  # Call the helper function to obtain regression metrics for training data
  results_train = regression_metrics(y_train, y_train_pred, verbose = verbose,
                                     output_dict=output_frame,
                                     label='Training Data')
  print()
  # Get predictions for test data
  y_test_pred = reg.predict(X_test)
  # Call the helper function to obtain regression metrics for test data
  results_test = regression_metrics(y_test, y_test_pred, verbose = verbose,
                                  output_dict=output_frame,
                                    label='Test Data' )
  # Store results in a dataframe if ouput_frame is True
  if output_frame:
    results_df = pd.DataFrame([results_train,results_test])
    # Set the label as the index
    results_df = results_df.set_index('Label')
    # Set index.name to none to get a cleaner looking result
    results_df.index.name=None
    # Return the dataframe
    return results_df.round(3)



In [ ]:
# Instantiate default random forest model
rf = RandomForestRegressor(random_state = 42)

# Model Pipeline
rf_pip = make_pipeline(col_transformer, rf)

In [ ]:
# Fit the model pipeline on the training data only
rf_pip.fit(X_train, y_train)


Pipeline(steps=[('columntransformer',
                 ColumnTransformer(transformers=[('numeric',
                                                  Pipeline(steps=[('simpleimputer',
                                                                   SimpleImputer()),
                                                                  ('standardscaler',
                                                                   StandardScaler())]),
                                                  Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
       'Outlet_Establishment_Year'],
      dtype='object')),
                                                 ('categorical',
                                                  Pipeline(steps=[('simpleimputer',
                                                                   SimpleImputer(fill_value='MISSING',
                                                                                 strategy='constant')),
                                                                  ('onehotencoder',
                                                                   OneHotEncoder(handle_unknown='ignore',
                                                                                 sparse_output=False))]),
                                                  Index(['Item_Fat_Content', 'Item_Type', 'Outlet_Identifier', 'Outlet_Size',
       'Outlet_Location_Type', 'Outlet_Type'],
      dtype='object'))],
                                   verbose_feature_names_out=False)),
                ('randomforestregressor',
                 RandomForestRegressor(random_state=42))])

In [ ]:
# Use custom function to evaluate default model
evaluate_regression(rf_pip, X_train, y_train, X_test, y_test)



------------------------------------------------------------
Regression Metrics: Training Data
------------------------------------------------------------
- MAE = 296.168
- MSE = 182,890.628
- RMSE = 427.657
- R^2 = 0.938

------------------------------------------------------------
Regression Metrics: Test Data
------------------------------------------------------------
- MAE = 765.934
- MSE = 1,214,345.911
- RMSE = 1,101.974
- R^2 = 0.560


#####The model is underfit because the R sqaured value for the test data is lower than that of the training data

#####The random forest model has better test scores than the linear regression model because its MAE is  lower than that of the linear regression model

In [ ]:
# Parameters for tuning
rf_pip.get_params()

{'memory': None,
 'steps': [('columntransformer',
   ColumnTransformer(transformers=[('numeric',
                                    Pipeline(steps=[('simpleimputer',
                                                     SimpleImputer()),
                                                    ('standardscaler',
                                                     StandardScaler())]),
                                    Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
          'Outlet_Establishment_Year'],
         dtype='object')),
                                   ('categorical',
                                    Pipeline(steps=[('simpleimputer',
                                                     SimpleImputer(fill_value='MISSING',
                                                                   strategy='constant')),
                                                    ('onehotencoder',
                                                     OneHotEncoder(handle_unknown='ignore',

In [ ]:
# Define param grid with options to try
params = {'randomforestregressor__max_depth': [None,10,15,20],
          'randomforestregressor__n_estimators':[10,100,150,200],
          'randomforestregressor__min_samples_leaf':[2,3,4],
          'randomforestregressor__max_features':['sqrt','log2',None],
          'randomforestregressor__oob_score':[True,False],
          }



In [ ]:
# Instantiate the gridsearch
gridsearch = GridSearchCV(rf_pip, params, n_jobs=-1, cv = 3, verbose=1)


In [ ]:
# Fit the gridsearch on training data
gridsearch.fit(X_train, y_train)

Fitting 3 folds for each of 288 candidates, totalling 864 fits


GridSearchCV(cv=3,
             estimator=Pipeline(steps=[('columntransformer',
                                        ColumnTransformer(transformers=[('numeric',
                                                                         Pipeline(steps=[('simpleimputer',
                                                                                          SimpleImputer()),
                                                                                         ('standardscaler',
                                                                                          StandardScaler())]),
                                                                         Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
       'Outlet_Establishment_Year'],
      dtype='object')),
                                                                        ('categorical',
                                                                         Pipeline(steps=[('simpleimputer',
                                                                                          SimpleImpu...
                                       ('randomforestregressor',
                                        RandomForestRegressor(random_state=42))]),
             n_jobs=-1,
             param_grid={'randomforestregressor__max_depth': [None, 10, 15, 20],
                         'randomforestregressor__max_features': ['sqrt', 'log2',
                                                                 None],
                         'randomforestregressor__min_samples_leaf': [2, 3, 4],
                         'randomforestregressor__n_estimators': [10, 100, 150,
                                                                 200],
                         'randomforestregressor__oob_score': [True, False]},
             verbose=1)

In [ ]:
# Obtain best parameters
gridsearch.best_params_

{'randomforestregressor__max_depth': 10,
 'randomforestregressor__max_features': None,
 'randomforestregressor__min_samples_leaf': 3,
 'randomforestregressor__n_estimators': 100,
 'randomforestregressor__oob_score': True}

In [ ]:
# Define and refit best model
best_rf = gridsearch.best_estimator_
evaluate_regression(best_rf, X_train, y_train, X_test, y_test)

------------------------------------------------------------
Regression Metrics: Training Data
------------------------------------------------------------
- MAE = 653.232
- MSE = 868,013.110
- RMSE = 931.672
- R^2 = 0.707

------------------------------------------------------------
Regression Metrics: Test Data
------------------------------------------------------------
- MAE = 734.426
- MSE = 1,118,362.595
- RMSE = 1,057.527
- R^2 = 0.595


#####The performance improved with the tuned model because it has a higher R squared value and its MAE is lower than that of the default model

##Evaluation Phase

#### I recommend the GridsearchCv model because it has the highest R squared value and a Lowest Mae value
- The model shows us 59.5% of the dependant variables variation with the target variable
-MAE
-I selected MAE because the lower its value is then the less errors that occur during the computation of the model
-The model is overfitting because the training data is better than the test data